File: python/refactoring/animals/zoo.py
from .animal import Animal
from .cow import Cow
from .dog import Dog
from .horse import Horse


def make_noise(animal: Animal) -> str:
    if isinstance(animal, Cow):
        return f"{animal.name}" + " says 'mooh'"
    if isinstance(animal, Dog):
        return f"{animal.name}" + " says 'woof'"
    if isinstance(animal, Horse):
        return f"{animal.name}" + " says 'neigh'"
    raise ValueError("unknown animal, unknown sound")

File: csharp/Refactoring/Animals/Zoo.cs
namespace Refactoring.Animals;

public class Zoo
{
    public static string MakeNoise(Animal animal)
    {
        if (typeof(Cow).IsInstanceOfType(animal))
        {
            return animal.Name + " says 'mooh'";
        }
        if (typeof(Dog).IsInstanceOfType(animal))
        {
            return animal.Name + " says 'woof'";
        }
        if (typeof(Horse).IsInstanceOfType(animal))
        {
            return animal.Name + " says 'neigh'";
        }
        throw new ArgumentException("unknown animal, unknown sound");
    }
}
